spark_client/services.py

```diff
@@ -870,3 +897,46 @@ def pyspark_shell(self, cli_property: Optional[str], extra_args: List[str]):
             self.logger.debug(submit_cmd)
             with environ(KUBECONFIG=self.kube_interface.kube_config_file):
                 os.system(submit_cmd)
+
+    def add_optional_detected_driver_host(self, conf: PropertyFile):
+        spark_driver_host = self.detect_host()
+        if spark_driver_host:
+            return PropertyFile({"spark.driver.host": spark_driver_host}) + conf
+        else:
+            return conf
+
+    def check_driver_host_extra_conf(self) -> bool:
+        args_including_conf, remaining_args = parse_arguments_with(
+            [add_logging_arguments, custom_parser, add_conf_arguments]
+        )
+
+        if args_including_conf.conf:
+            conf_args = dict(
+                conf_entry.split("=", 1) for conf_entry in args_including_conf.conf
+            )
+        else:
+            conf_args = {}
+
+        return "spark.driver.host" in conf_args
+
+    def detect_host(self) -> Any:
+        try:
+            host_py = socket.gethostbyname(socket.gethostname()).strip()
+        except Exception:
+            return None
+
+        try:
+            host_bash = (
+                subprocess.check_output(
+                    "hostname -I | cut -d' ' -f1", shell=True, stderr=None
+                )
+                .decode("utf-8")
+                .strip()
+            )
```

Comment:
I think logic can be more reliably (idea taken here - https://stackoverflow.com/a/166589)
* we have IP and port of kubernetes API in `self.service_account.api_server` var.
* spark client must be able to connect to it
* if we can connect to it, we can return our IP from TCP connection
```
import socket
host = self.service_account.api_server.split(":")[1].split("/")[-1]
port = self.service_account.api_server.split(":")[2] if len(api_server.split(":")) == 3 else 433

s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
s.connect((host, port))
print(s.getsockname()[0])
s.close()
```

Reply:
I really like this idea!